### ISSUE TITLE
Missing Symbols, no Dependency Errors?

### ISSUE THREAD

**Issue by buzzwick** [microsoft/navcontainerhelper#2985]
Running a self-hosted agent, build pipeline fails due to missing symbols from a dependency app, but no indication that the dependency app is either missing or not available. Log output from Compile segment:

2023-04-12T17:35:12.6024432Z ##[section]Starting: Compile app
2023-04-12T17:35:12.6331076Z ==============================================================================
2023-04-12T17:35:12.6331469Z Task         : PowerShell
2023-04-12T17:35:12.6331784Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2023-04-12T17:35:12.6332254Z Version      : 2.212.0
2023-04-12T17:35:12.6332511Z Author       : Microsoft Corporation
2023-04-12T17:35:12.6332869Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2023-04-12T17:35:12.6333258Z ==============================================================================
2023-04-12T17:35:14.1270490Z Generating script.
2023-04-12T17:35:14.1666964Z ========================== Starting Command Output ===========================
2023-04-12T17:35:14.1907256Z ##[command]"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agent\_work\_temp\ee87e9c6-3beb-4a7d-82dd-7f6bb53e6c8a.ps1'"
2023-04-12T17:35:15.0089103Z BcContainerHelper version 4.0.16
2023-04-12T17:35:15.5609721Z BC.HelperFunctions emits usage statistics telemetry to Microsoft
2023-04-12T17:35:23.7146135Z Using Symbols Folder: C:\agent\appfiles\
2023-04-12T17:35:53.5638775Z Processing dependency Microsoft_Application_19.0.0.0 ()
2023-04-12T17:35:53.5939886Z Dependency App exists
2023-04-12T17:35:53.6346351Z Processing dependency Microsoft_System_19.0.0.0 ()
2023-04-12T17:35:53.6352442Z Dependency App exists
2023-04-12T17:35:53.6410578Z Processing dependency CAI Software_CAi Software Precious Metals Solution_19.0.1.51 (8788f8f7-2afb-4cd8-9935-068947847f2a)
2023-04-12T17:35:53.6672580Z Dependency App exists
2023-04-12T17:35:53.9488925Z Compiling...
2023-04-12T17:35:54.0109315Z .\alc.exe /project:"C:\ProgramData\BcContainerHelper\Extensions\NobleMetalBuild\Source" /packagecachepath:"C:\shareddata" /out:"C:\ProgramData\BcContainerHelper\Extensions\NobleMetalBuild\Source\output\CAI Software_Noble Metal Services Custom Mods_19.0.0.54.app" /assemblyprobingpaths:"C:\Program Files (x86)\Microsoft Dynamics NAV\190\RoleTailored Client","C:\Program Files\Microsoft Dynamics NAV\190\Service","C:\Program Files (x86)\Open XML SDK\V2.5\lib","c:\Windows\Microsoft.NET\Assembly","C:\Test Assemblies\Mock Assemblies"
2023-04-12T17:36:17.0691180Z Microsoft (R) AL Compiler version 8.0.7.55163
2023-04-12T17:36:17.1492349Z Copyright (C) Microsoft Corporation. All rights reserved
2023-04-12T17:36:17.1581332Z 
2023-04-12T17:36:17.1598791Z Compilation started for project 'Noble Metal Services Custom Mods' containing '60' files at '13:35:55.304'.
2023-04-12T17:36:17.1683355Z 
2023-04-12T17:36:17.1975045Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/pageext/NOB100VendorContractHdrLst.PageExt.al(18,33): error AL0185: Table 'CAIRef Vendor Contract Header' is missing
2023-04-12T17:36:17.2028848Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/codeunit/NOB100CustomFunctions.Codeunit.al(8,24): error AL0185: Table 'CAiRef Refining Setup' is missing
2023-04-12T17:36:17.2050766Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/codeunit/NOB100CustomFunctions.Codeunit.al(13,58): error AL0185: Table 'CAiRef Lot Detail' is missing
2023-04-12T17:36:17.2059480Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/codeunit/NOB100CustomFunctions.Codeunit.al(15,24): error AL0185: Table 'CAiRef Refining Setup' is missing
2023-04-12T17:36:17.2078989Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/codeunit/NOB100CustomFunctions.Codeunit.al(17,16): error AL0185: Table 'CAiRef Lot' is missing
2023-04-12T17:36:17.2085085Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/codeunit/NOB100CustomFunctions.Codeunit.al(18,19): error AL0185: Codeunit 'CAiRef External System' is missing
2023-04-12T17:36:17.2090559Z ##[error]./C:/ProgramData/BcContainerHelper/Extensions/NobleMetalBuild/Source/src/codeunit/NOB100CustomFunctions.Codeunit.al(21,18): error AL0185: Table 'CAiRef Prod. Order Header' is missing
2023-04-12T17:36:17.2098838Z ##

...and so on, for every symbol reference in the app. I have confirmed that the dependency .app file is where it is supposed to be, in the Symbols folder. We have run this same compile many times, has something changed in BCContainerHelper?

**Comment by jonaswre** [comment]
Hi,
I had a similar issue when the app was present but the right artifact was not released/published/uploaded yet.

**Comment by buzzwick** [comment]
Hi!
Can you elaborate on what you mean by "right artifact was not released/published/uploaded'? Is that something we have to do, or is that supposed to happen during the build run?

**Comment by freddydk** [comment]
Do you know which app contains these missing symbols?
Assuming it is: CAI Software_CAi Software Precious Metals Solution_19.0.1.51 
Is it the right version that you have provided to the build?

**Comment by buzzwick** [comment]
That is the correct app for the symbols (we only have one), and that version is present in the appfiles folder. Can't show the screenshot here, you will have to take my word that it is and has been there throughout. 
We have been through this a few times, having to make sure our main .app file is there, so I can't figure why it would balk at this version. If you need more log data, let me know.

**Comment by jonaswre** [comment]
> Hi! Can you elaborate on what you mean by "right artifact was not released/published/uploaded'? Is that something we have to do, or is that supposed to happen during the build run?

Sorry, yes this error means you're trying to compile with the wrong symbols. For me (using github packages) this means the right version wasn't yet published to github packages.
Im not sure where you get your symbols from, do you just store them in a folder on the server? Im not sure how multiple versions of the same app in the folder might be handled? Maybe you can describe the mechanism you are using to manage your dependency in more detail? 

@freddydk maybe its time to recommend github packages as the default? It's been flawless for me for a while now, 36 apps published to AppSource are a good indicator for how well it works. I never think about dependency problems anymore. As I said for me the only way this error can occur is, if the right version wasn't published yet. And thats very straight forward to track down when using github packages.

**Comment by buzzwick** [comment]
Okay I'm going to need a little more background. On a self-hosted agent, we have a folder on the server where we place the .app file from which the necessary symbols (ours) are to be derived. We found we had to do the same thing--I.e., place an .app file in that folder--for the MS symbols: system, application, etc. Placing multiple versions of any of these in the same folder was not a problem until just recently (we have multiple versions of our main app codebase for reasons related to client deployments).  

No experience with GitHub packages--what they are, where they come from, where they go. But we are now in AL-Go for our base app. Perfectly happy to go with packages if someone wants to point us in the right direction, how to get started.
